---
title: "FLI1 patient manip 1 merge analysis"
author: "Laurent"
date: "11/2021"
always_allow_html: yes
output: 
  rmarkdown::html_document:
    code_folding: hide
    theme: cerulean
    toc: true
toc_float:
  toc_collapsed: true
pdf_document:
  toc: true
editor_options: 
  chunk_output_type: inline
---

---
title:  "Seurat merging on FLI1 D5 and D11 (Run 325 and 331)"
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir ="/home/hannouchel/workspace/FLI1/05_Output/03_Seurat/01_merge/01_Seurat_Output/01_FLI1_2021_set3_set5/RDS/Analysis_done")
```

## FLI1 merge analysis {.tabset .tabset-fade}

### Global analysis

- Loading Libraries 
```{r, include=FALSE}
# Load packages, data and functions

library(sctransform)
library(reticulate)
library(scigenex)

# Load all necessary packages
library(Seurat)
library(htmlwidgets)
library(DT)
library(plotly)
# library(DDRTree)
library(pheatmap)
library(reshape2)
library("gridExtra")
library(knitr)
library(scater)
# install.packages("amap")
# library(amap)
#library(kableExtra)
#library(monocle)
#library(scran)
library(dplyr)
library(ggrepel)
library(fmsb)
```

```{r env_loading, include=FALSE}
#  Path to the folder that will contain output objects

OUTPUT_PATH <- "/home/hannouchel/workspace/FLI1/05_Output/03_Seurat/01_merge/01_Seurat_Output/02_FLI1_Run325_331_J5_J11/RDS/Regress"
OUTPUT_PATH_DONE <- "/home/hannouchel/workspace/FLI1/05_Output/03_Seurat/01_merge/01_Seurat_Output/02_FLI1_Run325_331_J5_J11/RDS/Analysis_Done"
OUTPUT_PATH_DEGS <- "/home/hannouchel/workspace/FLI1/05_Output/03_Seurat/01_merge/01_Seurat_Output/02_FLI1_Run325_331_J5_J11/DEGs/INPUT_seurat"
OUTPUT_PATH_Radarplot <- "/home/hannouchel/workspace/FLI1/05_Output/03_Seurat/01_merge/01_Seurat_Output/02_FLI1_Run325_331_J5_J11/RadarPlot"

# Set the random number seed
set.seed(1234)
```

```{r}
FLI1_ctrl <- readRDS(paste(OUTPUT_PATH_DONE,"Ctrl/17_01_22_FLI1_ctrl_filtered_cell_cycle_regress_seurat_done_with_celltypes_2.rds", sep = "/"))
FLI1_pat <- readRDS(paste(OUTPUT_PATH_DONE,"Pat/12_12_22_FLI1_manip2_pat_filtered_cell_cycle_regress_seurat_done_with_celltypes_2.rds", sep= "/"))
```

- Merge 2 datasets (ctrl & pat)
```{r}
FLI1_manip2_combined <- merge(FLI1_ctrl, y = FLI1_pat, add.cell.ids = c("ctrl", "pat"), project = "FLI1")
# FLI1_manip2_combined@meta.data <- FLI1_manip2_combined@meta.data[,-c(20:21)]
table(FLI1_manip2_combined$HTO_souporcell_classification)
```

- Some quality controls: Count = reads; Features = genes
```{r}
Idents(FLI1_manip2_combined) <- FLI1_manip2_combined$HTO_souporcell_classification
hist(FLI1_manip2_combined@meta.data$nCount_RNA)
hist(FLI1_manip2_combined@meta.data$nFeature_RNA)
```

- Percentage of mitochondrial genes
```{r}
mito.genes <- grep("MT-", rownames(FLI1_manip2_combined@assays$RNA), value=T)
mito.genes
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
FLI1_manip2_combined[["percent.mt"]] <- PercentageFeatureSet(FLI1_manip2_combined, pattern = "MT-")

#Modif LJO : Compute mitochondrial percentage threshold
discard.mito=isOutlier(FLI1_manip2_combined[["percent.mt"]][,1],type="higher")
mito.threshold=min(FLI1_manip2_combined[["percent.mt"]][,1][discard.mito])

# Visualize QC metrics as a violin plot
VlnPlot(FLI1_manip2_combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3,pt.size=0)
```

- Ploting nFeature_RNA and mitochondrial percent against nCount_RNA
```{r}
plot1 <- FeatureScatter(FLI1_manip2_combined, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", pt.size = 0.5)
plot2 <- FeatureScatter(FLI1_manip2_combined, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0.5)
plot1
plot2
```

- Normalize data
```{r}
setwd(OUTPUT_PATH)

# standard log-normalization
FLI1_manip2_combined <- NormalizeData(FLI1_manip2_combined, normalization.method = "LogNormalize", scale.factor = 10000)

# write.table(FLI1_manip2_combined@assays$RNA@data,"log_norm_filtered_FLI1_manip2_combined.txt",
# 	    sep="\t", quote=F, col.names=NA)
```

- Some counts after normalization
```{r}
FLI1_manip2_combined
table(FLI1_manip2_combined$HTO_souporcell_classification)
```

- Choose top 1k variable genes
```{r}
FLI1_manip2_combined <- FindVariableFeatures(object=FLI1_manip2_combined, selection.method = "vst", mean.function = ExpMean, dispersion.function = LogVMR, binning.method = "equal_width", num.bin = 20, y.cutoff = 0.5, nfeatures = 1000, verbose=TRUE)
#write.table(icare@assays$RNA@var.features,
#	     paste("Top1000_variable_genes.",experiment,"_",params$sample,".txt", sep=""),
#	     sep="\t", quote=F, col.names=NA
#)

top40 <- head(VariableFeatures(FLI1_manip2_combined), 40)
top100 <- head(VariableFeatures(FLI1_manip2_combined), 100)
```

- Plot variable genes:
```{r Plot variable genes}
plot1 <- VariableFeaturePlot(FLI1_manip2_combined)
LabelPoints(plot = plot1, points = top100, repel = TRUE)
```

- Assign Cell-Cycle Scores
```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

FLI1_manip2_combined <- CellCycleScoring(FLI1_manip2_combined, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# view cell cycle scores and phase assignments
head(FLI1_manip2_combined[[]])
```

- Regress out cell cycle scores during data scaling
```{r}
# FLI1_manip2_combined <- ScaleData(FLI1_manip2_combined,  vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(FLI1_manip2_combined), display.progress = TRUE)
# setwd(OUTPUT_PATH)
# saveRDS(FLI1_manip2_combined,file = "09_01_23_FLI1_manip1_combined_regress_no_analysis.rds")
FLI1_manip2_combined_regress <- readRDS(paste(OUTPUT_PATH,"09_01_23_FLI1_manip1_combined_regress_no_analysis.rds",sep="/"))
```

- Running a PCA on cell cycle genes after correction
```{r}
FLI1_manip2_combined_regress <- RunPCA(FLI1_manip2_combined_regress, features = VariableFeatures(object = FLI1_manip2_combined_regress), npcs=40, pcs.print = 5, seed.use=42, rev.pca = FALSE)
DimPlot(FLI1_manip2_combined_regress,group.by="Phase", reduction = "pca")+ggtitle("PCA - Cell cycle")
DimPlot(FLI1_manip2_combined_regress, group.by = "HTO_souporcell_classification", reduction = "pca")
```

- JackStraw and ElbowPlot:
```{r}
FLI1_manip2_combined_regress <- JackStraw(FLI1_manip2_combined_regress, num.replicate = 100)
FLI1_manip2_combined_regress <- ScoreJackStraw(FLI1_manip2_combined_regress, dims = 1:20)
           
JackStrawPlot(FLI1_manip2_combined_regress, dims = 1:15)
           
ElbowPlot(FLI1_manip2_combined_regress)
```

- Perform clustering: 
```{r}
setwd(OUTPUT_PATH)
FLI1_manip2_combined_regress <- FindNeighbors(object = FLI1_manip2_combined_regress, k.param = 12, dims = 1:5, compute.SNN = TRUE, prune.SNN = 1/15)

FLI1_manip2_combined_regress <- FindClusters(object= FLI1_manip2_combined_regress, modularity.fxn = 1,
             resolution = 0.6, algorithm = 1, n.start = 10, n.iter = 10,
             random.seed = 42, temp.file.location = NULL, edge.file.name = NULL,
             verbose = TRUE)
clusters_res <- FLI1_manip2_combined_regress[["RNA_snn_res.0.6"]][,1]
names(clusters_res)=colnames(FLI1_manip2_combined_regress)

# write.table(clusters_res,
# 	    "clusters_res_0.8_FLI1_manip2_combined_regress.txt",
# 	    sep="\t", quote=F, col.names=NA)
```

- Look at PCA results again, with clusters
```{r}
DimPlot(FLI1_manip2_combined_regress, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(FLI1_manip2_combined_regress, dims=c(3, 4), reduction = "pca", pt.size=1)

```

- UMAP 
```{r}
setwd(OUTPUT_PATH)
FLI1_manip2_combined_regress <- RunUMAP(FLI1_manip2_combined_regress, dims = 1:10, umap.method="uwot", seed.use=10, n.components=2, n.neighbors = 10 , spread=1, min.dist=0.1)
DimPlot(FLI1_manip2_combined_regress, reduction = "umap", pt.size=0.7, label = TRUE)

# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
UMAP_coord <- FLI1_manip2_combined_regress@reductions$umap@cell.embeddings
# write.table(UMAP_coord,
# 	    "UMAP_coordinates_FLI1_manip2_combined_regress.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

- UMAP Cell cycling
```{r}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
 
FLI1_manip2_combined_regress <- CellCycleScoring(FLI1_manip2_combined_regress,s.features = s.genes, g2m.features = g2m.genes)
DimPlot(FLI1_manip2_combined_regress,group.by="Phase")+ggtitle("UMAP - Cell cycle")
```

- tSNE: 
```{r}
FLI1_manip2_combined_regress <- RunTSNE(object= FLI1_manip2_combined_regress,  dims = 1:10, dim.embed = 2, seed.use = 42, tsne.method="Rtsne", reduction="pca")
DimPlot(FLI1_manip2_combined_regress, reduction = "tsne", pt.size=0.7, label = TRUE)
tSNE_coord <- FLI1_manip2_combined_regress@reductions$tsne@cell.embeddings
# write.table(tSNE_coord,
# 	    "tSNE_coordinates_FLI1_manip2_combined_regress.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

- Top 50 marker genes by clusters and exclude ribosomal genes
```{r}
setwd(OUTPUT_PATH)

ribosomal_genes <- c("RPS17","RPS9","MRPS18B","RPS18","MRPS36","MRPS18B","RPS4Y2","RPS4Y1","RPS25","RPL7A","MRPS6","MRPS18B","RPS18","MRPS18B","RPS18","MRPL57","TBCE","RPS18","MRPS12","MRPS31","RPL3","MRPL23","RPL3L","RPS2","RPL17-C18orf32","RPS4X","RPS13","RPL36","RPS21","RPL21","RPL14","MRPS24","MRPL54","RPS24","RPS15A","RPS6","MRPS25","MRPL36","RPL36AL","MRPS17","MRPL16","MRPS30","FAU","MRPS16","RPS7","RPL23","RPS10","RPL10L","RPL35","RPL17","SRBD1","RPL22","RPL4","MRPS36","RPL37","MRPS2","RPL36A","RPL36A-HNRNPH2","RPL30","MRPL49","RPS19","RPL12","RPS5","RPS27L","RPS29","RPSA","MRPL33","MRPL1","RPS17","RPL10","MRPL13","MRPL19","MRPL15","MRPL41","RPS12","MRPS9","MRPS11","MRPL3","RPL31","RPS3","RPL38","RPL22L1","RPL39","MRPL46","RPL18A","RPL7","RPS3A","RPLP2","RPLP0","RPS16","MRPL23","MRPL22","MRPL18","RPL7L1","MRPL2","MRPS34","MRPS33","RPL28","RPL7A","MRPL4","RPL35A","RPL13","RPL24","RPS20","RPS15","MRPL14","MRPL21","MRPL32","NDUFA7","RPS28","RPL15","RPL18","RPL9","RPL37A","RPS11","MRPS23","MRPL52","MRPL43","MRPS18B","RPLP1","MRPS18A","RSL24D1","RPL27","RPL13A","MRPS14","MRPL47","MRPS35","RPL8","RPL26L1","MRPL11","RPL26","TBCE","RPS27AP5","MRPL34","MRPS12","MRPS5","MRPS18C","RPL29","RPL39L","MRPL28","MRPL51","RPL34","RPL10A","MRPL12","RPL32","RPS14","MRPL35","MRPL17","RPL11","RPS23","RPS9","RPL23A","RPS18","MRPL42","MRPL10","UBA52","DAP3","MRPS7","RPL19","RPS26","RPL41","MRPS22","RPL6","RPS8","TBCE","MRPL30","RPS27","RPL27A","MRPS21","MRPL27","MRPL20","MRPL55","MRPL37","RPS27A","MRPL24","RPL5","MRPL9","MRPS15")

all.genes <- rownames(FLI1_manip2_combined_regress)
ribosomal_genes_filtered <- ribosomal_genes[!duplicated(ribosomal_genes)]
all.genes_without_ribosomal = setdiff(all.genes, ribosomal_genes_filtered)
# length(ribosomal_genes)
# length(ribosomal_genes_filtered)
# length(all.genes)
# length(all.genes_without_ribosomal)
for(i in 0:(max(as.numeric(levels(FLI1_manip2_combined_regress@meta.data$seurat_clusters))))) {
  assign(paste("cluster", i,".markers", sep = ""), FindMarkers(FLI1_manip2_combined_regress, features = all.genes_without_ribosomal, ident.1 = i, test.use = "bimod", only.pos = TRUE))
    print(paste("cluster", i,".markers", sep = ""))
    print(head(get(paste("cluster", i,".markers", sep = "")),n=50))
#     write.table(get(paste("cluster", i,".markers", sep = "")),
# 		paste("cluster", i,"markers_FLI1_manip2_combined_regress.txt", sep = "_"),
# 		quote=FALSE,row.names = TRUE)
}
```

- FeaturePlot 
Dimensional reduction plot, with cells colored by a quantitative feature
```{r echo=FALSE}
# gene_list <- c("CRHBP","EMCN","HLF","AVP","SPINK2","CSF3R","GLIPR1","MIAT","IGHM","ITGA2B", "CD34","ELANE","AZU1","SERPINB10","MPO", "PRTN3", "CSF1R", "KLF1","TFR2","MPIG6B","GP9", "PF4", "SELP", "VWF","KIT")

gene_list <- c("THY1","PROM1","FLT3","CRHBP","HOPX","AVP","KIT","SPINK2","IGHM","GLIPR1","CSF3R","CPA3","CLC","PRG2","TPSAB1","RHEX","ALOX5AP","MS4A2","CD9","LMNA","CAVIN2","LAT","VWA5A","ITGA2B","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44","MPO","ELANE","PRTN3","AZU1","SERPINB10","CTSG","CSTA","RNASE3","CD38","TFRC","DEPTOR","KLF1","ACSM3","APOC1","TFR1","ANK1","TMSB10","ALDH1A1","MT-ATP8","H2AY","GYPB","CED","TFR2","HBB","APOC1","ANK1","ACSM3","BLVRB","CD9","STOM","LAT","PLEK","GP9","GP6","GP1BA","PF4","RASGRP2","MPIG6B","VWF","PPBP","LTBP1","SH3BP5","SELP","TMEM40","P2YR1","MYH10","MYH9","SLFN14","CD74","VIM","RNASE2","CSF2R","TUBB1","RAB27B","HPSE","FSCN1")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.6))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}

```

### Celltype signature (Supervised Method)

- Plot HSPC signature using 5 best candidate genes
```{r}
# list("THY1","PROM1","FLT3","CRHBP","HOPX","AVP","KIT","SPINK2","CSF3R","AVP","IGHM","GIPR")
tryCatch({
  FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("AVP","PROM1","CRHBP","FLT3","HOPX","IGHM","GLIPR1"),
name = "HSPC")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})

FeaturePlot(object = FLI1_manip2_combined_regress, features = "HSPC1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for HSPC signature
```{r}
gene_list <- c("AVP","PROM1","CRHBP","FLT3","HOPX","IGHM","GLIPR1")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot CMP signature using 5 best candidate genes
```{r}
# list("CSF3R","CPA3","CLC","PRG2","TPSAB1","RHEX","ALOX5AP","MS4A2")
tryCatch({
  FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("CPA3","PRG2","RHEX","ALOX5AP","MS4A2"),
name = "CMP")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "CMP1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for CMP signature
```{r}
gene_list <- c("CPA3","PRG2","RHEX","ALOX5AP","MS4A2")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot CMP-MK-primed signature using 5 best candidate genes
```{r}
# list("CD9","LMNA", "CAVIN2","LAT","VWA5A","ITGA2B","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44")
# list("VWA5A","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44","ITGA2B")
tryCatch({
  FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("VWA5A","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44"),
name = "CMP_MK_primed")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "CMP_MK_primed1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for CMP-MK-primed signature
```{r}
gene_list <- c("VWA5A","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot CMP-non-primed signature using 5 best candidate genes
```{r}
tryCatch({
  FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("MS4A3","RNASE2","CLC","CSF2RB","EPX","CTSG"),
name = "CMP_non_primed")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "CMP_non_primed1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for CMP-non-primed signature
```{r}
gene_list <- c("MS4A3","RNASE2","CLC","CSF2RB","EPX","CTSG")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot GMP signature using 5 best candidate genes
```{r}
# list("MPO","ELANE","PRTN3","AZU1","SERPINB10", "CTSG","CSTA","RNASE3")
# list("CSTA","MPO","ELANE","PRTN3","AZU1","IFT122","RNASE3")
# CED not present in genes dataset
tryCatch({
FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("CSTA","MPO","ELANE","PRTN3","AZU1","RNASE3"),
name = "GMP")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "GMP1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for GMP signature
```{r}
gene_list <- c("CSTA","MPO","ELANE","PRTN3","AZU1","RNASE3")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot MEP signature using 5 best candidate genes
```{r}
# list("CD38","TFRC","DEPTOR","KLF1","ACSM3","APOC1","ANK1","TMSB10","ALDH1A1","GYPB")
# list("CD38","KLF1","DEPTOR","ACSM3","APOC1","TFRC","ANK1","TMSB10","ALDH1A1","MT-ATP8","GYPB", "H2AY")
# list("KLRG1","IER2","KBTBD8","GASK1B","MINAR1","RAPH1","ZNF449","TNMD","TMEM98","MED17","PREX2","PER3","CELSR1","PIAS2","EMC2","KCNN4","ICAM5","DNASE2","AL021546,1","MRPL18","EXOC2","CNOT6","MSH3")
# list("CD38","KLF1","DEPTOR","ACSM3","APOC1","TFRC","ANK1","TMSB10","ALDH1A1","MT-ATP8","GYPB")

# H2AY =  MACROH2A1(not present) and TFR1 = TFRC
tryCatch({
FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("CD38","KLF1","DEPTOR","ACSM3","APOC1","ANK1","TMSB10","ALDH1A1","MT-ATP8","GYPB","TFRC"),name = "MEP")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "MEP1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for MEP signature
```{r}
gene_list <- c("CD38","KLF1","DEPTOR","ACSM3","APOC1","ANK1","TMSB10","ALDH1A1","MT-ATP8","GYPB","TFRC")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot MEP-ERP signature using 5 best candidate genes
```{r}
# list("TFR2","HBB","APOC1","ANK1","ACSM3","BLVRB")
# list("BLVRB","HBB","APOC1","TFR2")
# list("KLF1","AHSPP","RHAG")
# list("KLF1","AHSPP","RHAG", "APOC1","TFR2")
# list("KLF1","RHAG", "APOC1","TFR2")

tryCatch({
FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("HBB","BLVRB","APOC1","TFR2"),name = "MEP_ERP")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "MEP_ERP1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for MEP-ERP signature
```{r}
gene_list <- c("HBB","BLVRB","APOC1","TFR2")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot MEP-MKP signature using 5 best candidate genes
```{r}
# list("CD9","STOM","LAT","PLEK","ITGA2B")
# list("C3orf58","PDIA5","RGS18","PDGFA","PLEK","ARHGAP6")
tryCatch({
FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("PDIA5","RGS18","PDGFA","PLEK","ARHGAP6") ,name = "MEP_MKP")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "MEP_MKP1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for MEP-MkP signature
```{r}
gene_list <- c("PDIA5","RGS18","PDGFA","PLEK","ARHGAP6")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot MKP_MK signature using 5 best candidate genes
```{r}
# list("MPIG6B","PF4","GP9","VWF","PPBP","SH3BP5","SELP","LTBP1","TMEM40","GP6","GP1BA","RASGRP2")
# list("MPIG6B","PF4","GP9","VWF","PPBP","SH3BP5","SELP","LTBP1","TMEM40")
# list("PF4","GP6","VWF","TRPC6","ITGB3","FCER1G","SELP")
tryCatch({
FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("MPIG6B","PF4","GP9","VWF","PPBP","SH3BP5","SELP","LTBP1","TMEM40"),
name = "MKP_MK")
  }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
FeaturePlot(object = FLI1_manip2_combined_regress, features = "MKP_MK1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
```

- Looking individual gene expression for MkP_Mk signature
```{r}
gene_list <- c("MPIG6B","PF4","GP9","VWF","PPBP","SH3BP5","SELP","LTBP1","TMEM40")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- Plot MoMacro signature using 5 best candidate genes
```{r}
tryCatch({
FLI1_manip2_combined_regress <- AddModuleScore(FLI1_manip2_combined_regress, features = list("MERTK","SPP1","S100A8","S100A9","LILRB2","ITGAM","CD14","CD68","FCN1","VCAN"),
name = "MoMacro")

FeaturePlot(object = FLI1_manip2_combined_regress, features = "MoMacro1" ,reduction = "umap", order = TRUE,
pt.size = 0.6)
}, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
```

- Looking individual gene expression for MoMacro signature
```{r}
gene_list <- c("MERTK","SPP1","S100A8","S100A9","LILRB2","ITGAM","CD14","CD68","FCN1","VCAN","TLR4")

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = FLI1_manip2_combined_regress, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 0.5))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- DotPlot

* Look genes representative about a signature differenciation (supervised method)
```{r,fig.width=22,fig.height=8}

signature = list(HSPC = c("AVP","PROM1","CRHBP","FLT3","HOPX","IGHM","SPINK2","CSF3R"), #"GLIPR1"
                 CMP = c("CPA3","PRG2","RHEX","ALOX5AP","MS4A2","CLC","RNASE2","CSF2RB","CTSG"),
                 #primedCMP = c("VWA5A","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44"),
                 #NoprCMP = c("MS4A3","RNASE2","CLC","CSF2RB","EPX","CTSG"),
                 GMP =  c("MPO","ELANE","PRTN3","AZU1","RNASE3"), #"CSTA"
                 MP_M = c("CTSH","LYZ","SAMHD1","S100A8","S100A9","IGSF6","ITGAM","CD14","CD68","FCN1"),
                 MEP = c("CD38","KLF1","ALDH1A1","BLVRB","HBB","GYPB","MYH10","FSCN1"), #"MT-ATP8","DEPTOR","ACSM3","TMSB10","TFRC","APOC1","ANK1"
                 #ERP = c("HBB","BLVRB","APOC1","TFR2"),
                 MKP_MK = c("VWA5A","CAVIN2","LMNA","LAT","CD9","ITGA2B","MPIG6B","PF4","GP9","GP6","GP1BA","VWF","PPBP","SH3BP5","SELP","LTBP1","TMEM40","TUBB1","RAB27B","HPSE"))


FLI1_manip2_combined_regress_copy <- FLI1_manip2_combined_regress
# Define an order of cluster identities
my_levels <- c(13,12,8,4,1,0,7,6,5,3,2,14,10,11,9)

Idents(FLI1_manip2_combined_regress_copy) <- factor(Idents(FLI1_manip2_combined_regress_copy), levels= my_levels)

# clean les gènes en soublons par type cellulaires
un <- unlist(signature)
res <- Map(`[`, signature, relist(!duplicated(un), skeleton = signature))


dp <- DotPlot(FLI1_manip2_combined_regress_copy, features = res, dot.scale = 7) + theme(axis.text.x = element_text(size = 12), text = element_text(size = 15),legend.title = element_text(size=14), 
        legend.text = element_text(size=15)) + geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +  scale_colour_gradient2(low = "steelblue", mid = "ivory1", high = "red") + guides(size=guide_legend(override.aes=list(shape=21, colour="black", fill="white"))) + RotatedAxis() 
dp
```

- DotPlot 

* Personal signature without celltype
```{r,fig.width=15,fig.height=5}

my_levels <- c(13,12,8,4,1,0,7,6,5,3,2,14,10,11,9)

# res <- c("AVP","PROM1","CRHBP","FLT3","HOPX","IGHM","CD74","SPINK2","CSTA","VIM","TMSB10","MPO","ELANE","PRTN3","AZU1","RNASE3","CED","CPA3","PRG2","RHEX","ALOX5AP","MS4A2","VWA5A","TPSAB1","KIT","HPGDS","TPSB2","SAMSN1","CLC","CSF2RB","CTSG","RNASE2","TFRC","ALDH1A1","KLF1","CD38","DEPTOR","ANK1","GYPB","TFR2","HBB","BLVRB","APOC1","MPIG6B","PF4","GP9","VWF","PPBP","SH3BP5","SELP","TMEM40","CD9","SAT1","TUBB1","RAB27B","HPSE")

res <- c("AVP","PROM1","CRHBP","FLT3","HOPX","IGHM","SPINK2","CSF3R","CPA3","PRG2","RHEX","ALOX5AP","MS4A2","CLC","RNASE2","CSF2RB","CTSG","MPO","ELANE","PRTN3","AZU1","RNASE3","CD38","KLF1","ALDH1A1","BLVRB","HBB","GYPB","MYH10","FSCN1","VWA5A","CAVIN2","LMNA","LAT","CD9","ITGA2B","MPIG6B","PF4","GP9","GP6","GP1BA","VWF","PPBP","SH3BP5","SELP","LTBP1","TMEM40","TUBB1","RAB27B","HPSE")

dp <- DotPlot(FLI1_manip2_combined_regress_copy, features = res, dot.scale = 7) + theme(axis.text.x = element_text(size = 10), text = element_text(size = 14),legend.title = element_text(size=14), legend.text = element_text(size=15)) + geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +  scale_colour_gradient2(low = "steelblue", mid = "ivory1", high = "red") + guides(size=guide_legend(override.aes=list(shape=21, colour="black", fill="white"))) + RotatedAxis() 
dp
```

- DotPlot 

* Signature littérature
```{r, fig.width=15,fig.height=5}

signature = list(HSPC = c("THY1","PROM1","FLT3","CRHBP1","HOPX","AVP","KIT","SPINK2","CSF3R"),
                 CMP = c("CPA3","CLC","PRG2","TPSAB1"),
                 GMP =  c("MPO","ELANE","PRTN3","AZU1","SERPINB10","CTSG"),
                 MEP = c("CD38","TFRC","DEPTOR","KLF1","TFR2","HBB","APOC1","ANK1","ACSM3"),
                 ERP = c("HBB","APOC1","TFR2"),
                 MKP_MK = c("VWA5A","LMNA","CAVIN2","LAT","CD9","STOM","PLEK","ITGA22B","GP9","GP6","GP1BA","PF4","RASGRP2","MPIG6B","VWF","PPBP","SELP","LTBP"))
 
FLI1_manip2_combined_regress_copy <- FLI1_manip2_combined_regress
# Define an order of cluster identities
my_levels <- c(13,12,8,4,1,0,7,6,5,3,2,14,10,11,9)

Idents(FLI1_manip2_combined_regress_copy) <- factor(Idents(FLI1_manip2_combined_regress_copy), levels= my_levels)

# clean les gènes en doublons par type cellulaires
un <- unlist(signature)
res <- Map(`[`, signature, relist(!duplicated(un), skeleton = signature))


dp <- DotPlot(FLI1_manip2_combined_regress_copy, features = res, dot.scale = 7) + theme(axis.text.x = element_text(size = 10), text = element_text(size = 14),legend.title = element_text(size=14), 
        legend.text = element_text(size=15)) + geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +  scale_colour_gradient2(low = "steelblue", mid = "ivory1", high = "red") + guides(size=guide_legend(override.aes=list(shape=21, colour="black", fill="white"))) + RotatedAxis() 
dp
```

### Celltype signature (Unsupervised Method)

- Table: Top DE gene expression by cluster  
Checking most differentially expressed gene with FindAllMarkers in each clusters compared to all others.
```{r , echo=FALSE, message=FALSE, warning=FALSE}
# FindAllMarkers
Idents(FLI1_manip2_combined_regress) <- "RNA_snn_res.0.6"
markers <- FindAllMarkers(FLI1_manip2_combined_regress, features = all.genes_without_ribosomal,only.pos = T, test.use = "bimod", min.pct = 0.1, logfc.threshold = 0.2)

# Table
top5_genes <- markers %>% group_by(cluster) %>% top_n(5, avg_log2FC)
top20_genes <- markers %>% group_by(cluster) %>% top_n(20, avg_log2FC)

datatable(markers, filter = 'top', options = list(pageLength = 20)) %>% formatRound(2:4, 2) %>% formatSignif(c(1,5), digits = 2)
```

- Heatmap 
TOP 20 gene by cluster
```{r Heatmap_pat,  echo=FALSE,  fig.width=20,fig.height=15}
str_top_genes <- unique(as.character(top20_genes$gene))
DoHeatmap(object = FLI1_manip2_combined_regress, features = as.character(top20_genes$gene),group.by = "RNA_snn_res.0.6", group.bar = T, label = T,size = 3, draw.lines = T, lines.width = 10,group.bar.height = 0.02)+theme(text = element_text(size = 8))
```

- DotPlot 
TOP 20 gene by cluster (Unsupervised method)
```{r dotplot_pat, echo=FALSE, fig.width=15,fig.height=5}

FLI1_manip2_combined_regress_copy <- FLI1_manip2_combined_regress
# Define an order of cluster identities
my_levels <- c(13,12,8,4,1,0,7,6,5,3,2,14,10,11,9)
Idents(FLI1_manip2_combined_regress_copy) <- factor(Idents(FLI1_manip2_combined_regress_copy), levels= my_levels)

dp <- DotPlot(FLI1_manip2_combined_regress_copy, features = unique(top5_genes$gene), dot.scale = 7) + theme(axis.text.x = element_text(size = 12)) + geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5) +  scale_colour_gradient2(low = "steelblue", mid = "ivory1", high = "red") + guides(size=guide_legend(override.aes=list(shape=21, colour="black", fill="white"))) + RotatedAxis() 
dp
```

- Add few columns of interest: Status and Celltype with Status
```{r}
myTable <- FLI1_manip2_combined_regress@meta.data

for(i in 1:length(rownames(myTable))){
  if(myTable$HTO_souporcell_classification[i]%in% "Ctrl1-J5" | myTable$HTO_souporcell_classification[i]%in% "Ctrl1-J11" | myTable$HTO_souporcell_classification[i]%in% "Ctrl2-J5" | myTable$HTO_souporcell_classification[i]%in% "Ctrl2-J11"){
    myTable$FLI1status[i] <- "Ctrl"
  }
  if(myTable$HTO_souporcell_classification[i]%in% "FLI1-J5" | myTable$HTO_souporcell_classification[i]%in% "FLI1-J11"){
    myTable$FLI1status[i] <- "Pat"
    myTable$celltype_final[i] <- myTable$celltype[i]
  }
}

FLI1_manip2_combined_regress@meta.data <- myTable
```

- Création colonne celltype_ident
```{r}
FLI1_manip2_combined_regress@meta.data$celltype_ident <- paste(FLI1_manip2_combined_regress@meta.data$FLI1status,FLI1_manip2_combined_regress@meta.data$celltype_final, sep="_")
```

- ViolinPlot
Interest genes by clusters
```{r}
gene_list <- c("THY1","PROM1","FLT3","CRHBP","HOPX","AVP","KIT","SPINK2","IGHM","GLIPR1","CSF3R","CPA3","CLC","PRG2","TPSAB1","RHEX","ALOX5AP","MS4A2","CD9","LMNA","CAVIN2","LAT","VWA5A","ITGA2B","KRT1","TMEM176B","KIT","HPGDS","TPSB2","TPSAB1","SAMSN1","CD44","MPO","ELANE","PRTN3","AZU1","SERPINB10","CTSG","CSTA","RNASE3","CD38","TFRC","DEPTOR","KLF1","ACSM3","APOC1","TFR1","ANK1","TMSB10","ALDH1A1","MT-ATP8","H2AY","GYPB","CED","TFR2","HBB","APOC1","ANK1","ACSM3","BLVRB","CD9","STOM","LAT","PLEK","GP9","GP6","GP1BA","PF4","RASGRP2","MPIG6B","VWF","PPBP","LTBP1","SH3BP5","SELP","TMEM40","P2YR1","MYH10","MYH9","SLFN14","CD74","VIM","RNASE2","CSF2R","TUBB1","RAB27B","HPSE","FSCN1")

FLI1_manip2_combined_regress_copy <- FLI1_manip2_combined_regress
# Define an order of cluster identities
my_levels <- c(9,11,10,14,2,3,5,6,7,0,1,4,8,12,13)
Idents(FLI1_manip2_combined_regress_copy) <- factor(Idents(FLI1_manip2_combined_regress_copy), levels= my_levels)

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(VlnPlot(object = FLI1_manip2_combined_regress_copy, features = gene_list[i]))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- ViolinPlot 
nCount and nFeature by clusters
```{r}
VlnPlot(object = FLI1_manip2_combined_regress_copy, features = "nCount_RNA", pt.size = 0.1)
VlnPlot(object = FLI1_manip2_combined_regress_copy, features = "nFeature_RNA", pt.size = 0.1)
```

-ViolinPlot
Gene expression by Ctrl and Pat celltypes
```{r}
FLI1_manip2_combined_regress_copy <- FLI1_manip2_combined_regress
Idents(FLI1_manip2_combined_regress_copy) <- FLI1_manip2_combined_regress_copy@meta.data$celltype_ident
my_levels <- c("Ctrl_HSPC","Pat_HSPC","Ctrl_GMP","Pat_GMP","Ctrl_CMP","Pat_CMP","Ctrl_MkprimedCMP","Pat_MkprimedCMP","Ctrl_MP_M","Pat_MP_M","Ctrl_MEP","Pat_MEP","Ctrl_MkP_Mk","Pat_MkP_Mk")
Idents(FLI1_manip2_combined_regress_copy) <- factor(Idents(FLI1_manip2_combined_regress_copy), levels= my_levels)

for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(VlnPlot(object = FLI1_manip2_combined_regress_copy, features = gene_list[i], pt.size = 0.1))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

- ViolinPlot 
nCount and nFeature bby Ctrl and Pat celltypes
```{r}
VlnPlot(object = FLI1_manip2_combined_regress_copy, features = "nCount_RNA")
VlnPlot(object = FLI1_manip2_combined_regress_copy, features = "nFeature_RNA")
```

### Summary Plot

- Metadata construction
```{r}
setwd(OUTPUT_PATH)
metadata <- as.data.frame(FLI1_manip2_combined_regress@meta.data)

#Substitute "RNA_snn_res.0.5" with "RNA_snn_res"
#colnames(metadata)=gsub("RNA_snn_res.[0-9.]+$","RNA_snn_res",colnames(metadata))
metadata$UMAP_1 <- FLI1_manip2_combined_regress@reductions$umap@cell.embeddings[,1]
metadata$UMAP_2 <- FLI1_manip2_combined_regress@reductions$umap@cell.embeddings[,2]
metadata$tSNE_1 <- FLI1_manip2_combined_regress@reductions$tsne@cell.embeddings[,1]
metadata$tSNE_2 <- FLI1_manip2_combined_regress@reductions$tsne@cell.embeddings[,2]
metadata$PC1 <- FLI1_manip2_combined_regress@reductions[["pca"]]@cell.embeddings[,1]
metadata$PC2 <- FLI1_manip2_combined_regress@reductions[["pca"]]@cell.embeddings[,2]
metadata$sample_timepoint <- paste(metadata$timepoint, metadata$HTO_souporcell_classification, sep="_")

# write.table(metadata,
# 	    "metadata_SLFN14_combined_pat_cluster_filtered.txt",
# 	    sep='\t', quote=F, col.names=NA)

```

- Plot nFeature_RNA 
The genes/feature number per cell
```{r fig.width=18,fig.height=10}
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=nFeature_RNA), size=0.4)+ scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))
```

- Plot nCount_RNA 
The reads/nCount number per cell
```{r fig.width=18,fig.height=10}
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=nCount_RNA), size=0.4)+ scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))
```

- Ggplot function
```{r}
get_cluster_centroids <- function(comp_1,comp_2) {
	coords=cbind(comp_1,comp_2)
	clusters=clusters_res
	centers=c()
	for (cl in unique(clusters)) {
		sel=clusters==cl
		x_center=mean(coords[sel,1])
		y_center=mean(coords[sel,2])
		centers=rbind(centers,c(cl,x_center,y_center))
	}
	centers=data.frame("Cluster"=as.factor(centers[,1]),
			   "x_center"=as.numeric(centers[,2]),
			   "y_center"=as.numeric(centers[,3])
			   )
}

color_ordered2=c("indianred1", "darkorange3", "olivedrab3", "red4", "palegreen", "royalblue4", "palevioletred3", "orchid3", "olivedrab2", "palegreen3", "lightseagreen", "steelblue2", "royalblue2", "turquoise1", "chocolate1", "firebrick1", "magenta", "mediumorchid1","darkgreen", "indianred2", "blue")
```

- Plot Ctrl et Pat nFeature et nCount par Celltype
```{r fig.width=18,fig.height=10}

clusters_res <- metadata$RNA_snn_res.0.6
centers=get_cluster_centroids(metadata$UMAP_1,metadata$UMAP_2)

metadata_ctrl <- metadata[which(metadata$FLI1status == "Ctrl"),]

for(i in sort(unique(metadata_ctrl$celltype))){
  metadata2 <- metadata_ctrl[which(metadata_ctrl$celltype == i),]
  p1 <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2))+ geom_point(aes(colour=nFeature_RNA), size=0.5) + scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(paste("Ctrl",i,"nFeatureRNA", sep="_")) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,size=3)
  p2 <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2))+ geom_point(aes(colour=nCount_RNA), size=0.5) + scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(paste("Ctrl",i,"nCountRNA", sep="_")) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,size=3)
  plot(p1)
  plot(p2)
}

metadata_pat <- metadata[which(metadata$FLI1status == "Pat"),]

for(i in sort(unique(metadata_pat$celltype))){
  metadata2 <- metadata_pat[which(metadata_pat$celltype == i),]
  p1 <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2))+ geom_point(aes(colour=nFeature_RNA), size=0.5) + scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(paste("Pat",i,"nFeatureRNA", sep="_")) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,size=3)
  p2 <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2))+ geom_point(aes(colour=nCount_RNA), size=0.5) + scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(paste("Pat",i,"nCountRNA", sep="_")) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,size=3)
  plot(p1)
  plot(p2)
}
```

- Ggplot for all samples 
```{r fig.width=18,fig.height=10}

p <- ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.4) + scale_colour_manual(values=color_ordered2) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,size=3)

for(i in sort(unique(metadata$HTO_souporcell_classification))){
  metadata2 <- metadata[which(metadata$HTO_souporcell_classification == i),]
  clusters_res <- metadata2$RNA_snn_res.0.6
  p <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.4) + scale_colour_manual(values=color_ordered2) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(i) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)
  plot(p)
}
```

- Plot by status
```{r fig.width=15,fig.height=7}
# my_levels <- c("LT","MkP_Mk","MEP","MP_M","MkprimedCMP","CMP","GMP","HSPC")
my_levels <- c("MkP_Mk","MEP","MP_M","MkprimedCMP","CMP","GMP","HSPC")
DimPlot(FLI1_manip2_combined_regress, group.by = "celltype", split.by = "FLI1status", order = my_levels, label=T, label.size = 3.5)
```

- Plot by celltype
```{r fig.width=15,fig.height=10}
DimPlot(FLI1_manip2_combined_regress,  group.by = "celltype", label = T, order = my_levels, label.size = 3.5, pt.size = 0.8)
```

- Zoom for each Celltype
```{r,fig.width=15,fig.height=12}
df <- data.frame(FLI1_manip2_combined_regress_copy@meta.data,
                 UMAP_1 = FLI1_manip2_combined_regress_copy@reductions$umap@cell.embeddings[,1],
                 UMAP_2 = FLI1_manip2_combined_regress_copy@reductions$umap@cell.embeddings[,2]
                 )
for(i in unique(metadata$celltype)){
  df_celltype <- df[which(df$celltype == i),]
  df_celltype_ctrl <- df_celltype[df_celltype$FLI1status == "Ctrl",]
  df_celltype_pat <- df_celltype[df_celltype$FLI1status == "Pat",]

  p<- ggplot(df,
       aes(x=UMAP_1,y=UMAP_2),
       label = rownames(df))+
  geom_point(alpha=0.5,
             color="lightgrey",
             size = 0.5)+
  geom_point(data = df_celltype_ctrl,
             aes(x=UMAP_1,y=UMAP_2),
             color="salmon",
             size = 0.7)+
  geom_point(data = df_celltype_pat,
             aes(x=UMAP_1,y=UMAP_2),
             color="cornflowerblue",
             size = 0.7)+ theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(i) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,size=3)
  plot(p)
}
```

- Plot Ctrl vs Pat
```{r,fig.width=15,fig.height=10}
 DimPlot(FLI1_manip2_combined_regress, group.by = "FLI1status", pt.size = 0.7)
```

### DEGs Patient vs Control

- Patient vs Temoins for HSPC cluster
```{r}
setwd(OUTPUT_PATH_DEGS)
Idents(FLI1_manip2_combined_regress) <- FLI1_manip2_combined_regress$celltype_ident
HSPC_Patients_vs_Temoins.markers <- FindMarkers(FLI1_manip2_combined_regress, ident.1 = c("Pat_HSPC"), ident.2 = c("Ctrl_HSPC"), min.pct = 0.1, test.use = "bimod", only.pos=FALSE)
write.table(HSPC_Patients_vs_Temoins.markers, "DEGs_HSPC_Patients_vs_Temoins.markers.txt", sep="\t", quote=F)
```

- Patient vs Temoins for GMP cluster
```{r}
setwd(OUTPUT_PATH_DEGS)
Idents(FLI1_manip2_combined_regress) <- FLI1_manip2_combined_regress$celltype_ident
GMP_Patients_vs_Temoins.markers <- FindMarkers(FLI1_manip2_combined_regress, ident.1 = c("Pat_GMP"), ident.2 = c("Ctrl_GMP"), min.pct = 0.1, test.use = "bimod", only.pos=FALSE)
write.table(GMP_Patients_vs_Temoins.markers, "DEGs_GMP_Patients_vs_Temoins.markers.txt", sep="\t", quote=F)
```

- Patient vs Temoins for CMP cluster
```{r}
setwd(OUTPUT_PATH_DEGS)
Idents(FLI1_manip2_combined_regress) <- FLI1_manip2_combined_regress$celltype_ident
CMP_Patients_vs_Temoins.markers <- FindMarkers(FLI1_manip2_combined_regress, ident.1 = c("Pat_CMP"), ident.2 = c("Ctrl_CMP"), min.pct = 0.1, test.use = "bimod", only.pos=FALSE)
write.table(CMP_Patients_vs_Temoins.markers, "DEGs_CMP_Patients_vs_Temoins.markers.txt", sep="\t", quote=F)
```

- Patient vs Temoins for CMPprimed cluster
```{r}
setwd(OUTPUT_PATH_DEGS)
Idents(FLI1_manip2_combined_regress) <- FLI1_manip2_combined_regress$celltype_ident
CMPprimed_Patients_vs_Temoins.markers <- FindMarkers(FLI1_manip2_combined_regress, ident.1 = c("Pat_MkprimedCMP"), ident.2 = c("Ctrl_MkprimedCMP"), min.pct = 0.1, test.use = "bimod", only.pos=FALSE)
write.table(CMPprimed_Patients_vs_Temoins.markers, "DEGs_MkprimedCMP_Patients_vs_Temoins.markers.txt", sep="\t", quote=F)
```

- Patient vs Temoins for MEP cluster
```{r}
setwd(OUTPUT_PATH_DEGS)
Idents(FLI1_manip2_combined_regress) <- FLI1_manip2_combined_regress$celltype_ident
MEP_Patients_vs_Temoins.markers <- FindMarkers(FLI1_manip2_combined_regress, ident.1 = c("Pat_MEP"), ident.2 = c("Ctrl_MEP"), min.pct = 0.1, test.use = "bimod", only.pos=FALSE)
write.table(MEP_Patients_vs_Temoins.markers, "DEGs_MEP_Patients_vs_Temoins.markers.txt", sep="\t", quote=F)
```

- Patient vs Temoins for MkP_Mk cluster
```{r}
setwd(OUTPUT_PATH_DEGS)
Idents(FLI1_manip2_combined_regress) <- FLI1_manip2_combined_regress$celltype_ident
MkP_Mk_Patients_vs_Temoins.markers <- FindMarkers(FLI1_manip2_combined_regress, ident.1 = c("Pat_MkP_Mk"), ident.2 = c("Ctrl_MkP_Mk"), min.pct = 0.1, test.use = "bimod", only.pos=FALSE)
write.table(MkP_Mk_Patients_vs_Temoins.markers, "DEGs_MkP_Mk_Patients_vs_Temoins.markers.txt", sep="\t", quote=F)
```

- Volcano Plot
```{r}
# OUTPUT_PATH_Volcano <- paste(OUTPUT_PATH_DEGS,"temp2/",sep="/")

# setwd(paste(OUTPUT_PATH_DEGS,"temp/",sep="/"))

setwd(OUTPUT_PATH_DEGS)

DEGs_list <- list("DEGs_MkprimedCMP_Patients_vs_Temoins.markers.txt",
                  "DEGs_CMP_Patients_vs_Temoins.markers.txt",
                  "DEGs_GMP_Patients_vs_Temoins.markers.txt",
                  "DEGs_HSPC_Patients_vs_Temoins.markers.txt",
                  "DEGs_MEP_Patients_vs_Temoins.markers.txt",
                  "DEGs_MkP_Mk_Patients_vs_Temoins.markers.txt"
                  )


for (i in DEGs_list){
  DEGs_temp <-  read.table(i ,sep="\t")
  DEGs_temp$significant <- ifelse(DEGs_temp$p_val_adj < 0.05 & DEGs_temp$avg_log2FC > 0.37 | DEGs_temp$p_val_adj < 0.05 & DEGs_temp$avg_log2FC < -0.37,
                              ifelse(DEGs_temp$avg_log2FC > 0,"Upregulated_bimod_0.37","Downregulated_bimod_0.37"),"Not_Significative")
  # DEGs_temp$significant2 <- ifelse(DEGs_temp$p_val_adj < 0.05&DEGs_temp$avg_log2FC > 1 | DEGs_HSPC_Patients_vs_Temoins.markers_volcano$p_val_adj < 0.05 & DEGs_temp$avg_log2FC < -1,
  #                             ifelse(DEGs_temp$avg_log2FC > 0,"Upregulated_bimod_1","Downregulated_bimod_1"),"Not_Significative")
  fileName <- strsplit(i, split=".markers.txt")[[1]][1]
  write.table(DEGs_temp, file= paste(OUTPUT_PATH_DEGS,"/",fileName,"_volcano.txt",sep = ""), sep="\t", row.names=T, quote=F)
}

DEGs_volcano_list <- list("DEGs_MkprimedCMP_Patients_vs_Temoins_volcano.txt",
                  "DEGs_CMP_Patients_vs_Temoins_volcano.txt",
                  "DEGs_GMP_Patients_vs_Temoins_volcano.txt",
                  "DEGs_HSPC_Patients_vs_Temoins_volcano.txt",
                  "DEGs_MEP_Patients_vs_Temoins_volcano.txt",
                  "DEGs_MkP_Mk_Patients_vs_Temoins_volcano.txt"
                 )

# setwd(OUTPUT_PATH_Volcano)

for (i in DEGs_volcano_list){
  DEGs_temp_volcano <-  read.table(i, sep="\t")
  DEGs_temp_volcano$genename <- rownames(DEGs_temp_volcano)
  # number of genes upregulated (> 1.3 FC)
  length(rownames(DEGs_temp_volcano[DEGs_temp_volcano$significant == "Upregulated_bimod_0.37",]))
  # number of genes downregulated (< -1.3 FC)
  length(rownames(DEGs_temp_volcano[DEGs_temp_volcano$significant == "Downregulated_bimod_0.37",]))
  # number of genes upregulated (> 2 FC)
  # length(rownames(DEGs_temp_volcano[DEGs_temp_volcano$significant2 == "Upregulated_bimod_1",]))
  # number of genes downregulated (< -2 FC)
  # length(rownames(DEGs_temp_volcano[DEGs_temp_volcano$significant2 == "Downregulated_bimod_1",]))
  fileName <- strsplit(i, split="_volcano.txt")[[1]][1]
  p <- ggplot(DEGs_temp_volcano, aes(x = avg_log2FC, y = -log10(p_val_adj))) +
  geom_point(aes(color = significant))+
  scale_color_manual(values = c("blue", "lightgrey","red")) +
  theme_bw(base_size = 12) + theme(legend.position = "bottom") +
  scale_x_continuous( limits=c(-2.5, 2.5)) +
  # scale_y_continuous( limits=c(-1, 35)) +
  geom_vline(xintercept = c(-0.37, 0.37),10) +
  # geom_vline(xintercept = c(-1, 1),10) +
  geom_hline(yintercept = 1.3) +
  ggtitle(label = paste(fileName,"_Volcano_Plot")) +
  geom_text_repel(
    data = subset(DEGs_temp_volcano, avg_log2FC > 0.37 & p_val_adj < 0.05 | p_val_adj < 0.05 & avg_log2FC < -0.37),
    aes(label = genename),
    size = 3.4,
    box.padding = unit(0.25, "lines"),
    point.padding = unit(0.25, "lines"))
                 
  # png(paste(OUTPUT_PATH_Volcano,"Figure_Volcano/",fileName,"_Volcano_Plot",sep=""), width = 12, height = 6, units = 'in', res = 300)
  plot(p)
  # dev.off()
}
```

- Radar Plot

Matrice
```{r}
setwd(OUTPUT_PATH_Radarplot)
# Mise en place d'un tableau avec le nombre de cellules par échantillon et type cellulaire

myTable <- FLI1_manip2_combined_regress@meta.data
myTable$FLI1status_timepoint <- paste(myTable$timepoint, myTable$FLI1status, sep = "_")
myTable$samples_timepoint <- paste(myTable$timepoint, myTable$HTO_souporcell_classification, sep = "_")

# On instancie une matrice ayant un nombre de lignes égale au nombre d'individus et de colonne égale au nombre de types cellulaires
myMatrix <- matrix(nrow=length(unique(myTable$samples_timepoint)), ncol = length(unique(myTable$celltype_final)))
colnames(myMatrix) <-c("HSPC","GMP","CMP","MkprimedCMP","MP_M","MEP","MkP_Mk") 
rownames(myMatrix) <- unique((myTable$samples_timepoint))

# On rempli la matrice
for (y in unique(myTable$celltype_final)){
  for(i in unique(myTable$samples_timepoint)){
    myMatrix[i,y] <- nrow(myTable[which(myTable$samples_timepoint %in% i & myTable$celltype_final %in% y),])
  }
}  

# On instancie une matrice min max afin de calculer le min max pour chaque type cellulaire
min_max_Matrix <- matrix(nrow=2, ncol = length(unique(myTable$celltype_final)))
colnames(min_max_Matrix) <- c("HSPC","GMP","CMP","MkprimedCMP","MP_M","MEP","MkP_Mk") 
rownames(min_max_Matrix) <-c("Max","Min")
min_max_Matrix[1,]<- colMaxs(myMatrix)
min_max_Matrix[2,]<- colMins(myMatrix)

# On merge les 2 matrices qu'on convertit en data frame
myMatrix <- rbind(min_max_Matrix, myMatrix)

# On enregistre la matrice

write.table(myMatrix, paste(OUTPUT_PATH_Radarplot, "Celltypes_Statistics_nbCells_FLI1_samples_timepoint.tab", sep = "/"), sep = "\t", quote = FALSE)


######################### Percentage #############################

# On instancie une matrice ayant un nombre de lignes égale au nombre d'individus et de colonne égale au nombre de types cellulaires
myMatrix_percent <- matrix(nrow=length(unique(myTable$samples_timepoint)), ncol = length(unique(myTable$celltype_final)))
colnames(myMatrix_percent) <- c("HSPC","GMP","CMP","MkprimedCMP","MP_M","MEP","MkP_Mk") 
rownames(myMatrix_percent) <- unique((myTable$samples_timepoint))

# On rempli la matrice
for (y in unique(myTable$celltype_final)){
  for(i in unique(myTable$samples_timepoint)){
    myMatrix_percent[i,y] <- round(x = nrow(myTable[which(myTable$samples_timepoint %in% i & myTable$celltype_final %in% y),])*100/nrow(myTable[which(myTable$samples_timepoint %in% i),]), digit = 2)
  }
}  

# test <- data.frame(t(table(SLFN14_ctrl@meta.data$HTO_souporcell_classification)))
# test$Percent <- test$Freq / sum(test$Freq) * 100
# 
# test <- data.frame(t(table(SLFN14_ctrl@meta.data$celltype_final[SLFN14_ctrl@meta.data$HTO_souporcell_classification %in% "Ctrl2"])))
# test$Percent <- test$Freq / sum(test$Freq) * 100

# On instancie une matrice min max afin de calculer le min max pour chaque type cellulaire
min_max_Matrix_percent <- matrix(nrow=2, ncol = length(unique(myTable$celltype_final)))
colnames(min_max_Matrix_percent) <- c("HSPC","GMP","CMP","MkprimedCMP","MP_M","MEP","MkP_Mk") 
rownames(min_max_Matrix_percent) <-c("Max","Min")
min_max_Matrix_percent[1,]<- colMaxs(myMatrix_percent)
min_max_Matrix_percent[2,]<- colMins(myMatrix_percent)

# On merge les 2 matrices qu'on convertit en data frame
myMatrix_percent <- rbind(min_max_Matrix_percent, myMatrix_percent)

# On enregistre la matrice

write.table(myMatrix_percent, paste(OUTPUT_PATH_Radarplot, "Celltypes_Percentage_nbCells_FLI1_samples_timepoint.tab", sep = "/"), sep = "\t", quote = FALSE)

######################### Percentage with Ctrl pool and SLFN14 (No Zoom) ####################

# On instancie une matrice ayant un nombre de ligne égale au nombre d'individus et de colonne égale au nombre de types cellulaires
myMatrix_percent_sample_pool <- matrix(nrow=length(unique(myTable$FLI1status_timepoint)), ncol = length(unique(myTable$celltype_final)))
colnames(myMatrix_percent_sample_pool) <- c("HSPC","GMP","CMP","MkprimedCMP","MP_M","MEP","MkP_Mk")
rownames(myMatrix_percent_sample_pool) <- unique((myTable$FLI1status_timepoint))


# On rempli la matrice

for (y in unique(myTable$celltype_final)){
  for(i in unique(myTable$FLI1status_timepoint)){
      myMatrix_percent_sample_pool[i,y] <- round(x = nrow(myTable[which(myTable$FLI1status_timepoint %in% i & myTable$celltype_final %in% y),])*100/nrow(myTable[which(myTable$FLI1status_timepoint %in% i),]), digit = 2)
  }
}  

# On instancie une matrice min max afin de calculer le min max pour chaque type cellulaire (No Zoom)
min_max_Matrix_percent_sample_pool <- matrix(nrow=2, ncol = length(unique(myTable$celltype_final)))
colnames(min_max_Matrix_percent_sample_pool) <- c("HSPC","GMP","CMP","MkprimedCMP","MP_M","MEP","MkP_Mk")
rownames(min_max_Matrix_percent_sample_pool) <-c("Max","Min")
min_max_Matrix_percent_sample_pool[1,]<- colMaxs(myMatrix_percent_sample_pool)
min_max_Matrix_percent_sample_pool[2,]<- colMins(myMatrix_percent_sample_pool)

# On merge les 2 matrices qu'on convertit en data frame
myMatrix_percent_sample_pool <- rbind(min_max_Matrix_percent_sample_pool, myMatrix_percent_sample_pool)

# On enregistre la matrice

setwd(OUTPUT_PATH_Radarplot)
write.table(myMatrix_percent_sample_pool, paste(OUTPUT_PATH_Radarplot, "Celltypes_Percentage_nbCells_FLI1_samples_timepoint_pool.tab", sep = "/"), sep = "\t", quote = FALSE)
```

- Save RDS
```{r}
setwd(OUTPUT_PATH)
saveRDS(FLI1_manip2_combined_regress , file= "FLI1_manip2_combined_regress_seurat_done_2.rds")
```

#
